### ISSUE TITLE
Tests (and, consequently, my Yesod app) fail with conduit-0.4.1

### ISSUE THREAD

**Issue by cgaebel** [prowdsponsor/crypto-conduit#6]
sinkHash and sinkHmac are causing your test suite to fail:

http://pastebin.com/BGN80Pxm

Let me know if you need more information. I think this might be due to the change in conduit-0.4.1 which added a Finalizer type, however, I'm not familiar enough with it to tell.

**Comment by meteficha** [comment]
That pastebin doesn't work for me, would you mind copy & pasting here or putting on another pastebin?

My dev environment still haven't caught up with the new `conduit` version so it may take a few days for me to look into this issue.  But I'm usually very fast in merging pull requests ;-).

**Comment by cgaebel** [comment]
http://hpaste.org/raw/66925

And your code is pulled into the latest yesod, and prevents static files from compiling. It's kind of very bad.

I know absolutely nothing about your code or the latest conduit changes, so I wouldn't be comfortable making the change myself.

**Comment by meteficha** [comment]
Looks pretty bad =/.  I know that it's used by Yesod. @snoyberg was the one who did the changes last time, but he's AFK until at least Sunday.  Could you include your `ghc-pkg dump` as well?

When you say that your app fails, do you mean that your app really fails, or do you mean that the build process of your app runs `crypto-conduit`'s test suite and consequently fails?

**Comment by cgaebel** [comment]
I'm using cabal-dev. Here's the list of all the packages installed with your library:

http://hpaste.org/66934

I mean the build of my app calls crypto-conduit in some template haskell (I believe this is present in the scaffolding, even), and that fails.

**Comment by league** [comment]
I can reproduce, just posted to yesod's mailing list about this -- it prevents the scaffolded site from building because it tries to hash files in the `static/` subdir. I have only the vaguest understanding of conduits (from reading up over the past hour), but the issue must originate from here, in `Crypto/Conduit.hs:83`:

``` Haskell
      sink = sinkState C.initialCtx
                       push
                       (const $ fail "sinkHash")
```

where `sinkState` takes an initial state, a push function and a close function. From this, it seems that if it ever calls the close function (which it apparently does), then we get the "sinkHash" failure. I don't know how to reconcile that with the `StateDone` in the definition of `push`:

``` Haskell

      push ctx (Full bs) =
          let !ctx' = C.updateCtx ctx bs
          in return (StateProcessing ctx')
      push ctx (LastOne bs) =
          let !ret = C.finalize ctx bs
          in return (StateDone Nothing ret)
```

Clearly that expects to return the final hash result, but the `close :: state -> m output` parameter has its opportunity too, and that's where `fail "sinkHash"` happens.

**Comment by league** [comment]
Progress: I discovered that the following simple changes make the tests succeed again, with conduit-0.4.1. I don't know why the `close` function returns an output value if it is ignored, but the real result is taken from `StateDone`.

``` diff
diff --git a/src/Crypto/Conduit.hs b/src/Crypto/Conduit.hs
index b1af5e3..21fb490 100644
--- a/src/Crypto/Conduit.hs
+++ b/src/Crypto/Conduit.hs
@@ -82,7 +82,7 @@ sinkHash = blocked AnyMultiple blockSize =$ sink
     where
       sink = sinkState C.initialCtx
                        push
-                       (const $ fail "sinkHash")
+                       (const $ return undefined)

       push ctx (Full bs) =
           let !ctx' = C.updateCtx ctx bs
@@ -133,7 +133,7 @@ sinkHmac (C.MacKey key) = blocked AnyMultiple blockSize =$ sink

       sink = sinkState (C.updateCtx C.initialCtx ki)
                        push
-                       (const $ fail "sinkHmac")
+                       (const $ return undefined)

       push ctx (Full bs) =
           let !ctx' = C.updateCtx ctx bs
@@ -348,7 +348,7 @@ sinkCbcMac k = blocked StrictBlockSize blockSize =$ sink
           | B.null input = return (StateDone Nothing iv)
           | otherwise    = fail "sinkCbcMac: input has an incomplete final block."

-      close _ = fail "sinkCbcMac"
+      close _ = return undefined

       blockSize = (C.blockSize .::. k) `div` 8

```

**Comment by cgaebel** [comment]
You probably want a <code>return $ error "sinkHmac"</code>, but that looks fine to me. Anything will do at this point, as long as it works.

**Comment by league** [comment]
Yeah that makes sense, I was briefly forgetting the distinction between error and fail. This trick will pass the test suite, but unfortunately it's still not enough for yesod. You can get past the yesod build error with `staticFiles` but the first HTTP request bombs with the error (`undefined` in my case, but then I changed it to `error "foo"` like you suggested, and that's what shows up).

**Comment by snoyberg** [comment]
This is caused by too much strictness in conduit 0.4.1. I was trying to get around a space leak caused by too much laziness in GHC (see snoyberg/conduit#40 and snoyberg/conduit#41), but after some testing right now, I can avoid the space leak without the strictness annotations. I'll release a new version of conduit fairly soon.

I'm also going to have a look at crypto-conduit and see if some of this functions can be optimized based on our new infrastructure in conduit.

**Comment by snoyberg** [comment]
I've released conduit 0.4.1.1 which should hopefully solve the bug. Can anyone else confirm? I've also sent pull request #7, which includes a more efficient implementation of some of these functions, which also avoids the need to use `fail` as much.

**Comment by league** [comment]
I've confirmed that the changes in pull request 7 and conduit-0.4.1.1 work together, both for the crypto-conduit test suite (which is **much, much** faster) and in yesod-1.0.0.2. Thanks!!

**Comment by cgaebel** [comment]
Can this be merged in ASAP? I don't like having to maintain a fork of crypto-conduit just for my (and apparently, everyone else's) yesod app.

**Comment by snoyberg** [comment]
The new version of conduit is on Hackage, so you should be able to build against the most recent versions of both conduit and crypto-conduit and have a working system (I just did so yesterday in fact). The changes in #7 are just an optimization.

**Comment by meteficha** [comment]
#7 has been merged and released as 0.3.1.  Sorry for not taking enough care of this issue but I've been AFK in the last few days for personal reasons.  And thanks to everyone who hunted down this issue =).
